File: alexa_test.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
)

type alexaTestOutput struct {
	Domain     string `json:"domain"`
	Rank       int    `json:"rank"`
	StatusCode int    // `json:"statusCode"`
	Error      string `json:"Error,omitempty"`
}

var alexaTests = []struct {
	it         string
	endpoint   string
	errorText  string
	statusCode int
}{
	{"not domain(empty)", "/v1/alexa/get", "ERROR Bad Request, incorrect parameters", 400},
	{"wrong domain", "/v1/alexa/wrong", "ERROR Bad Request, incorrect parameters", 400},
	{"wrong domain", "/v1/alexa/too/much/parameters", "ERROR Bad Request, incorrect parameters", 400},
	{"not a valid domain name", "/v1/alexa/get/code%%tabs.com", `parse /v1/alexa/get/code%%tabs.com: invalid URL escape "%%t"`, 0},
	{"domain in routes parameters", "/v1/alexa/get/google.com", "", 200},
	{"www.github.com", "/v1/alexa/get/www.github.com", "", 200},
	{"github.com", "/v1/alexa/get/github.com", "", 200},
	{"valid domain not in alexa top 1m", "/v1/alexa/get/not-top-domain.com", "not-top-domain.com not in alexa top 1 million", 400},
	{"awwwards.com ,www in the middle of the domain", "/v1/alexa/get/awwwards.com", "", 200},
	{"progresswww.nl ,www at the end of the domain", "/v1/alexa/get/progresswww.nl", "", 200},
	{"www.gov.uk ,www as subdomain", "/v1/alexa/get/www.gov.uk", "", 200},
	{"gov.uk ,multiple results", "/v1/alexa/get/gov.uk", "", 200},
	{"www.nic.in ,www as subdomain", "/v1/alexa/get/www.nic.in", "", 200},
	{"bih.nic.in ,not www subdomain in multiple results", "/v1/alexa/get/bih.nic.in", "", 200},
	{"nic.in ,multiple results", "/v1/alexa/get/nic.in", "", 200},
}

func TestAlexaApi(t *testing.T) {
	c.App.Mode = "test"

	for _, test := range alexaTests {
		var to = alexaTestOutput{}
		pass := true
		//fmt.Println(`Test...`, test.endpoint, "...", test.it)
		req, err := http.NewRequest("GET", test.endpoint, nil)
		if err != nil {
			//fmt.Println(`------------------------------`)
			//fmt.Println(err.Error())
			//fmt.Println(test.errorText)
			//fmt.Println(`------------------------------`)
			if err.Error() != test.errorText {
				t.Errorf("Error Request %s\n", err.Error())
			} else {
				pass = false
			}
		}
		if pass {
			rr := httptest.NewRecorder()
			handler := http.HandlerFunc(router)
			handler.ServeHTTP(rr, req)
			if rr.Code != test.statusCode {
				t.Errorf("%s got %v want %v\n", test.endpoint, rr.Code, test.statusCode)
			}
			_ = json.Unmarshal(rr.Body.Bytes(), &to)
			if to.Error != test.errorText {
				t.Errorf("%s got %v want %v\n", test.endpoint, to.Error, test.errorText)
			}
		}
		fmt.Printf("Test OK...%s\n", test.endpoint)
	}
}

File: server/test/alexa.test.js
/* */

import { use, expect } from "chai";
import chaiHttp from "chai-http";
const chai = use(chaiHttp);

import { aux } from "../middlewares.js";
const url = "http://localhost:3000";

describe('ALEXA TEST ', function () {
  before(function (done) {
    done();
  });
  it('not domain (empty)', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('Error', 'Domain is empty');
        done();
      });
  });
  it('not valid parameters', function (done) {
    const t = "Bad request : /v1/alexa/&(%/&%%/ is not a valid url";
    chai.request.execute(url)
      .get('/v1/alexa/&(%/&%%/')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('msg', aux.badRequest);
        done();
      });
  });
  it('not a valid domain name', function (done) {
    const t = 'code%%tabs.com not in alexa top 1 million';
    chai.request.execute(url)
      .get('/v1/alexa?web=code%%tabs.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('Error', t);
        done();
      });
  });
  it('not a registered domain name', function (done) {
    const t = 'sure-this-is-not-registered.non not in alexa top 1 million';
    chai.request.execute(url)
      .get('/v1/alexa?web=sure-this-is-not-registered.non')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('Error', t);
        done();
      });
  });
  it('domain in query string', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa')
      .query({ web: 'google.com' })
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body.rank).to.be.oneOf([1, 2, 3, 4, 5]);
        expect(res.body).to.have.property('domain', 'google.com');
        expect(res.body).to.have.property('rank');
        done();
      });
  });
  it('domain in route parameters ', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=google.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body.rank).to.be.oneOf([1, 2, 3, 4, 5]);
        expect(res.body).to.have.property('domain', 'google.com');
        expect(res.body).to.have.property('rank');
        done();
      });
  });
  it('www.github.com', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=www.github.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'github.com');
        done();
      });
  });
  it('github.com', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=github.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'github.com');
        done();
      });
  });
  it('valid domain not in alexa top 1m', function (done) {
    const t = 'not-top-domain.com not in alexa top 1 million';
    chai.request.execute(url)
      .get('/v1/alexa?web=not-top-domain.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('Error', t);
        done();
      });
  });
  it('awwwards.com ,www in the middle of the domain', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=awwwards.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'awwwards.com');
        done();
      });
  });
  it('progresswww.nl ,www at the end of the domain', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=progresswww.nl')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'progresswww.nl');
        done();
      });
  });
  it('www.gov.uk ,www as subdomain', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=www.gov.uk')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'www.gov.uk');
        done();
      });
  });
  it('www.nic.in ,www as subdomain', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=www.nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'www.nic.in');
        done();
      });
  });
  it('bih.nic.in ,not www subdomain in a multiple results', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=bih.nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'bih.nic.in');
        done();
      });
  });
  it('nic.in , exists with www ', function (done) {
    chai.request.execute(url)
      .get('/v1/alexa?web=nic.in')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(200);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('domain', 'www.nic.in');
        done();
      });
  });
  it('params instead query 1', function (done) {
    const t = "Domain is empty";
    chai.request.execute(url)
      .get('/v1/alexa/codetabs.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('msg', aux.badRequest);
        done();
      });
  });
  it('params instead query 2', function (done) {
    const t = "Domain is empty";
    chai.request.execute(url)
      .get('/v1/alexa/get/codetabs.com')
      .end(function (err, res) {
        // console.log('PATH=> ', res.req.path)
        expect(err).to.be.null;
        expect(res).to.have.status(400);
        expect(res).to.be.json;
        expect(res.body).to.be.an('object');
        expect(res.body).to.have.property('msg', aux.badRequest);
        done();
      });
  });
});
